### PR TITLE
MoSoReadingListKit

### PR DESCRIPTION
**Pull request by CMasterson**
## Summary
A reading list component that uses Apollo.GraphQL to pull a users Pocket Saves list and displays it for consumption, sharing and archival.

## Implementation
Dependencies: Apollo v1.7.0, MoSoCore, MoSoAnalytics, DesignKit

### Temp code
Auth credentials are currently fetched by a transitory login flow through the LoginView. It writes the data to Keychain using code pulled from Pocket V8 for retrieval later. Later, we expect the MoSo auth token to be valid for Pocket transactions.

### General info
RLK is instantiated by creating a ReadingListView instance and passing it a ReadingListModel object.

ReadingListModel is responsible for holding the ViewModels, figuring out when it is time to fetch more data and passing back results to the view. It has some requirements:
- A Session Provider: a reference to a function with signature `() throws -> Session` (To simplify things, a typealias ReadingListSessionProvider exists). This function is expected to return a valid ReadingListSession object (token: String, guid: String) or throw a MoSoSessionError.
- A consumerKey: The same (for now, at least) as used in PocketV8. Required for the Pocket backend to talk to you.
- An Analytics Tracker: Must conform to the ReadingListTracker protocol.

PocketAccessLayer is responsible for interacting with the Pocket backend. It holds the consumerKey and SessionProvider objects. It keeps track of the current pagination state and when requested fetches the next page.

The view owns a LazyVStack of ReadingListItemCells that is bound to the ReadingListModel.readingListItems array.

MoSoSessionManager (renamed from MoSoSession) is responsible for holding the current MoSoUser (Perhaps multiple users in the near future) and holds the implementation of the SessionProvider logic mentioned above. Sessions themselves have been moved to individual MoSoUser instances to leave a path for multiple logins in the future.

## TODO

- Not all errors are routed to useful user messages.
- Polish of the UI is needed
- Opening the ShareSheet and tapping on the main body of an item closes the sheet, as expected but also opens that item's URL.
- Fetching is unidirectional, the Pagination state only increments. You can currently scroll for ever and ever. If it ever happens that we purge some of the user's earlier downloaded items (to conserve memory) we currently have no way of getting them back, outside of restarting the fetch from page 0.

### CODE REVIEW

**Comment by CMasterson** [comment]
There's some commented out code I can address.
Dave helped me get the Keychain stuff working and we talked about GroupID and if we need it or not. 
For now we decided to leave it out, we can add it later and it will currently only log a user out. We didn't feel like it was a good idea to try to guess at what a GroupID might look like without more context.

File: Demo/MozillaSocial-iOS/MoSoContent/MoSoContent.entitlements
Comment (dskuza) [inline Demo/MozillaSocial-iOS/MoSoContent/MoSoContent.entitlements:1]: Is this file still required if there are no entitlements added?

File: Demo/MozillaSocial-iOS/MoSoContent/PocketAuthenticationService.swift
Comment (dskuza) [inline Demo/MozillaSocial-iOS/MoSoContent/PocketAuthenticationService.swift:1]: This is pretty much a carry-over from the Pocket codebase, right?

File: Demo/MozillaSocial-iOS/apollo-ios-cli
Comment (dskuza) [inline Demo/MozillaSocial-iOS/apollo-ios-cli:1]: This is probably nice-to-have, but in the Pocket codebase this isn't committed. Just to note.

File: Sources/MoSoCore/MoSoUser.swift [hunk 1]
```diff
@@ -6,9 +6,14 @@
 public class MoSoUser {
     public let username: String
     public let identifier: String
+    @KeychainStorage public var session: MoSoSession?
+    @KeychainStorage public var pocketSession: MoSoSession?
 
-    public init(username: String, identifier: String) {
+    public init(username: String, identifier: String, session: MoSoSession? = nil) {
```
Comment (dskuza) [inline Sources/MoSoCore/MoSoUser.swift]: Is the added argument still required? Looks like session is initialized below.

File: Sources/ReadingListKit/GraphQL/Apollo Extensions.swift [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+enum ReadingListEndpointError: Error {
+    case couldNotLoadValidEndpointURL
+}
+
+extension ApolloClient {
+    static func createDefault(
+        sessionProvider: @escaping ReadingListSessionProvider,
+        consumerKey: String
+    ) -> ApolloClient {
+        guard let url = URL(string: ApolloClient.endpointURLString()) else { fatalError() }
```
Comment (dskuza) [inline Sources/ReadingListKit/GraphQL/Apollo Extensions.swift:17]: Looks like there's an endpoint error above that's not used. This function could throw, instead. Though, I know this is probably a carryover from the Pocket codebase.

File: Sources/ReadingListKit/GraphQL/AuthParamsInterceptor.swift [hunk 3]
```diff
@@ -0,0 +1,57 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+import Apollo
+
+class AuthParamsInterceptor: ApolloInterceptor {
+    var id: String
```
Comment (dskuza) [inline Sources/ReadingListKit/GraphQL/AuthParamsInterceptor.swift:9]: I don't think this is a necessary variable anymore.

File: Sources/ReadingListKit/GraphQL/Apollo Extensions.swift [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+        let authParams = AuthParamsInterceptor(
+            sessionProvider: sessionProvider,
+            consumerKey: consumerKey
+        )
+        let store = ApolloStore()
+        let interceptorProvider = PrependingUnauthorizedInterceptorProvider(
```
Comment (dskuza) [inline Sources/ReadingListKit/GraphQL/Apollo Extensions.swift:24]: Doesn't seem like the `PrependingUnauthorizedInterceptorProvider` is being used. Optional to remove it.

File: Sources/ReadingListKit/GraphQL/UnauthorizedInterceptor.swift [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        completion: @escaping (Result<Apollo.GraphQLResult<Operation.Data>, Error>) -> Void
+    ) where Operation: ApolloAPI.GraphQLOperation {
+        // This case will be sent from a ResponseCodeInterceptor, which is a part of the base DefaultInterceptorProvider
+        // that is used by our PrependingUnauthorizedInterceptorProvider. A 401 (Unauthorized) or 403 (Forbidden)  status code
+        // will cause this error to be handled. We can capture it, and post a notification  to then log a user out.
+        let invalidResponseCodes = [401, 403]
+        if case ResponseCodeInterceptor.ResponseCodeError.invalidResponseCode(response: let errorResponse, rawData: _) = error,
+           let statusCode = errorResponse?.statusCode,
+           invalidResponseCodes.contains(statusCode) {
+//            NotificationCenter.default.post(name: .unauthorizedResponse, object: nil)
```
Comment (dskuza) [inline Sources/ReadingListKit/GraphQL/UnauthorizedInterceptor.swift:46]: If we're keeping this type around, doesn't hurt to uncomment this. Otherwise, this type doesn't really do much.

File: Sources/ReadingListKit/PocketAccessLayerImplementation.swift [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+    func fetchSaves()
+    func initApolloClient()
+    func getItemForURL(_ urlString: String) async throws -> PocketGraph.ItemByURLQuery.Data.ItemByUrl
+    func archive(item: String)
+}
+
+public typealias ReadingListSessionProvider = () throws -> ReadingListSession
+
+enum PALError: Error {
```
Comment (dskuza) [inline Sources/ReadingListKit/PocketAccessLayerImplementation.swift]: It might be a bit longer, but imo `PocketAccessLayerError` is a bit more descriptive when reading. Same with `PALState`

File: Sources/ReadingListKit/PocketAccessLayerImplementation.swift [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+    var token: String { get }
+    var guid: String { get }
+}
+
+protocol PocketAccessLayer {
+    var delegate: ReadingListModelDelegate? { get set }
+
+    func fetchSaves()
+    func initApolloClient()
+    func getItemForURL(_ urlString: String) async throws -> PocketGraph.ItemByURLQuery.Data.ItemByUrl
```
Comment (dskuza) [inline Sources/ReadingListKit/PocketAccessLayerImplementation.swift]: I see that the usage of this function is in the reading list model, via `readingListItem`. However, I think it might be a good idea to abstract away the GraphQL usage into something else, like a struct, that models the object we'd like to utilize throughout the codebase. The current object is _very_ large, and contains a lot of data that we have to parse through at the view model layer. If we use the access layer as the source-of-truth of a simplified model that fits the overall use case better, that would help me, at least, better understand what we have to work with. It can always be expanded later.

File: Sources/ReadingListKit/PocketAccessLayerImplementation.swift [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+}
+
+enum PALState {
+    case idle
+    case fetching
+}
+
+class PocketAccessLayerImplementation: PocketAccessLayer {
+    let consumerKey: String
+    let pageSize: GraphQLNullable<Int> = GraphQLNullable<Int>(integerLiteral: 20)
```
Comment (dskuza) [inline Sources/ReadingListKit/PocketAccessLayerImplementation.swift]: Can you update the access control of some of these properties to `private` if they're not being used outside of this class? Just a little cleaner way of ensuring our public-facing API "appears" minimal, especially since GraphQL is an implementation detail here.

File: Sources/ReadingListKit/ReadingListModel.swift [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+    var pocketAccessLayer: PocketAccessLayer
+    var totalNumberOfItemsInReadingList: Int?
+    @Published private(set) var displayMode: ReadingListDisplayMode = .normal
+    @Published private(set) var readingListItems: [ReadingListCellViewModel] = []
+
+    public init(sessionProvider: @escaping ReadingListSessionProvider, consumerKey: String, analyticsTracker: ReadingListTracker) {
+        self.analytics = analyticsTracker
+        pocketAccessLayer = PocketAccessLayerImplementation(sessionProvider, consumerKey)
+        pocketAccessLayer.delegate = self
+        pocketAccessLayer.initApolloClient()
```
Comment (dskuza) [inline Sources/ReadingListKit/ReadingListModel.swift]: What's the reason behind not having this client just initialized within the access layer's `init`?

File: Sources/ReadingListKit/ReadingListModel.swift [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+
+        urlStrings.forEach { urlString in
+            Task {
+                guard let item = try? await pocketAccessLayer.getItemForURL(urlString) else { return }
+                let readingListItem = readingListItem(from: item)
+                await MainActor.run {
+                    readingListItems.append(readingListItem)
+                }
+            }
+        }
```
Comment (dskuza) [inline Sources/ReadingListKit/ReadingListModel.swift]: I wonder, in terms of rendering, if `forEach` here would cause more re-renders than just updating `readingListItems` in bulk. Perhaps room for performance improvement…?

File: Sources/ReadingListKit/ReadingListModel.swift [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+
+    // MARK: - Archive Items
+
+    func archive(item: String) {
+        pocketAccessLayer.archive(item: item)
+    }
+
+    // MARK: - ReadingListItem Helpers
+
+    func readingListItem(from item: PocketGraph.ItemByURLQuery.Data.ItemByUrl) -> ReadingListCellViewModel {
```
Comment (dskuza) [inline Sources/ReadingListKit/ReadingListModel.swift]: Same as another comment - I think that the implementation detail of items being backed by the Pocket graph should be abstracted away.

File: Sources/MoSoCore/MoSoSessionManager.swift [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+
+    public func loggedIn(_ user: MoSoUser) {
+        self.user = user
+    }
+
+    public func loggedOut() {
+        user = nil
+    }
+
+    public func activeSession() throws -> MoSoSession {
```
Comment (dskuza) [inline Sources/MoSoCore/MoSoSessionManager.swift:24]: Could you give insight into the benefits of throwing here rather than just returning an optional?

File: Sources/ReadingListKit/PocketAccessLayerImplementation.swift [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+    }
+
+    // MARK: - Fetch Saves
+
+    func fetchSaves() {
+        if state == .fetching { return }
+        state = .fetching
+
+        let query = PocketGraph.FetchSavesQuery(
+            pagination: .some(pagination),
```
Comment (dskuza) [inline Sources/ReadingListKit/PocketAccessLayerImplementation.swift]: I'm not too fond of keeping state here, and having `fetchSaves` implicitly load the next page. It feels to me like this type / layer shouldn't be really tracking state, and is rather just the "fetcher" of data. With that being said, I'd consider maybe returning the _cursor_ along with the items, such that when we fetch, we can fetch for items _after_ a cursor of a given size, with the cursor being the "state" that is captured (perhaps by a view model…?) as the "last" loaded page.

File: Sources/ReadingListKit/ReadingListModel.swift (hunk 10 above)
Comment (dskuza) [inline Sources/ReadingListKit/ReadingListModel.swift]: I'm also not sure this is sorting the items properly. I'm _assuming_ we're ordering newest -> oldest. If that's the case, then these items should be sorted by `createdAt`, descending.

File: Sources/ReadingListKit/ReadingListModel.swift [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+        let defaultImageURLString = "https://helios-i.mashable.com/imagery/articles/05fACELrEVc4kAfNQbhhcVh/hero-image.fill.size_1248x702.v1667556469.png"
+
+        let id = item.remoteID
+        let title = item.title ?? item.givenUrl
+        let subtitle = subtitle(for: item)
+        let contentURL = item.resolvedUrl ?? item.givenUrl
+        let thumbnailURL = image(for: item) ?? defaultImageURLString
+
+        return ReadingListCellViewModel(id: id, title: title, subtitle: subtitle, contentURL: contentURL, thumbnailURL: thumbnailURL)
```
Comment (dskuza) [inline Sources/ReadingListKit/ReadingListModel.swift]: I know we've had discussions on extensions before, but this sort of logic is what feels like would be a good candidate for extending ReadingListCellViewModel to have an initializer that takes an item type, and generates its correct properties there. This seems like we're expanding this view model into what the list is and what the individual rows should render. Just some room for separation, imo.

File: Sources/ReadingListKit/ReadingListModel.swift [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+    private let analytics: ReadingListTracker
+
+    var pocketAccessLayer: PocketAccessLayer
+    var totalNumberOfItemsInReadingList: Int?
+    @Published private(set) var displayMode: ReadingListDisplayMode = .normal
+    @Published private(set) var readingListItems: [ReadingListCellViewModel] = []
+
+    public init(sessionProvider: @escaping ReadingListSessionProvider, consumerKey: String, analyticsTracker: ReadingListTracker) {
+        self.analytics = analyticsTracker
+        pocketAccessLayer = PocketAccessLayerImplementation(sessionProvider, consumerKey)
```
Comment (dskuza) [inline Sources/ReadingListKit/ReadingListModel.swift]: Using a concrete implementation of PAL might make this class more difficult to test in the long run.

**Review by dskuza — COMMENTED** [review]
So, I think this is a good first pass. Thematically, I think there are a few changes:
1. Slim down our GraphQL files to _just_ what we need to use. Carrying over Pocket's files brings a lot of bloat to a much simplified experience.
2. Abstract away the usage of GraphQL types when passing around types across layers. I.e PAL should abstract away GraphQL, and maybe something fitting the DAO pattern would fit nicely, here.
3. 👀 PAL generally
4. Styling / structure nitpicks

Initially, I was thinking that maybe there would be a way to join the two `AuthenticationService` types into something reusable, but I think _for this_, they're just different enough where maybe right now that doesn't make sense. E.g Mastodon requires some nested calls just to get the values to provide to an OAuth flow. Just leaving this as a note to any other reviewers.

I think these are enough comments to make a first pass at handling review comments, then we can continue further.

File: Demo/MozillaSocial-iOS/MoSoContent/PocketAuthenticationService.swift
Comment (CMasterson) [reply Demo/MozillaSocial-iOS/MoSoContent/PocketAuthenticationService.swift:1]: Pretty much, yeah.
https://github.com/Pocket/pocket-ios/blob/develop/PocketKit/Sources/PocketKit/Authorization/AuthorizationClient.swift

File: Demo/MozillaSocial-iOS/apollo-ios-cli
Comment (CMasterson) [reply Demo/MozillaSocial-iOS/apollo-ios-cli:1]: How do we feel about removing generated files from the project and using Build Phases to create them from the declarations?
Glean is supposed to do that right now, and I'd like to see if Apollo can do that too.

File: Sources/MoSoCore/MoSoUser.swift (hunk 1 above)
Comment (CMasterson) [reply Sources/MoSoCore/MoSoUser.swift]: Removed.

File: Sources/MoSoCore/MoSoSessionManager.swift (hunk 12 above)
Comment (CMasterson) [reply Sources/MoSoCore/MoSoSessionManager.swift:24]: In general: I like throwing because returning nil often hides an error. Throws make it very explicit what is happening.

In this case we may want to have different flows based on if the user needs to refresh their token (which we may be able to do automatically) and there is no user logged in, which will require us to launch the login flow.

File: Sources/ReadingListKit/PocketAccessLayerImplementation.swift (hunk 13 above)
Comment (CMasterson) [reply Sources/ReadingListKit/PocketAccessLayerImplementation.swift]: This is a great suggestion, it fixes a few of my issues with this logic.

File: Sources/ReadingListKit/GraphQL/AuthParamsInterceptor.swift (hunk 3 above)
Comment (CMasterson) [reply Sources/ReadingListKit/GraphQL/AuthParamsInterceptor.swift:9]: It's declared inside Apollo in the `ApolloInterceptor` protocol.
I can't edit that to remove it as a requirement.

I'll change "Test" to "" for now.

File: Sources/ReadingListKit/GraphQL/Apollo Extensions.swift (hunk 4 above)
Comment (CMasterson) [reply Sources/ReadingListKit/GraphQL/Apollo Extensions.swift:24]: From my reading this is what is applying the authParams interceptor.
We need that, right?

File: Sources/ReadingListKit/GraphQL/UnauthorizedInterceptor.swift (hunk 5 above)
Comment (CMasterson) [reply Sources/ReadingListKit/GraphQL/UnauthorizedInterceptor.swift:46]: Does this do anything for us?
Can we remove the entire interceptor?

File: Sources/ReadingListKit/PocketAccessLayerImplementation.swift (hunk 7 above)
Comment (CMasterson) [reply Sources/ReadingListKit/PocketAccessLayerImplementation.swift]: How about a protocol that only reveals the properties we care about?

File: Sources/ReadingListKit/ReadingListModel.swift (hunk 10 above)
Comment (CMasterson) [reply Sources/ReadingListKit/ReadingListModel.swift]: We could probably .map the [urls] into [viewModel] and then append that once.

File: Sources/ReadingListKit/PocketAccessLayerImplementation.swift (hunk 7 above)
Comment (CMasterson) [reply Sources/ReadingListKit/PocketAccessLayerImplementation.swift]: Check out the updated code and see what you think.

File: Sources/ReadingListKit/ReadingListModel.swift (hunk 10 above)
Comment (CMasterson) [reply Sources/ReadingListKit/ReadingListModel.swift]: Updated!

File: Sources/ReadingListKit/ReadingListModel.swift (hunk 11 above)
Comment (CMasterson) [reply Sources/ReadingListKit/ReadingListModel.swift]: This got fixed* in the other changes you referenced.


*Fixed may mean not fixed, that's for later.

File: Sources/ReadingListKit/PocketAccessLayerImplementation.swift (hunk 13 above)
Comment (CMasterson) [reply Sources/ReadingListKit/PocketAccessLayerImplementation.swift]: Updated, the model now holds a String? that is the end of the last fetched page.
This will cause some knock-on changes to be needed in the Notification PR. But this is much better.

File: Sources/ReadingListKit/ReadingListModel.swift (hunk 14 above)
Comment (CMasterson) [reply Sources/ReadingListKit/ReadingListModel.swift]: While I like my ViewModels to be as dumb as possible, I did move this code to an extension of the GraphQL ItemByUrl type. So we kinda both get what we want?

File: Sources/ReadingListKit/ReadingListModel.swift (hunk 15 above)
Comment (CMasterson) [reply Sources/ReadingListKit/ReadingListModel.swift]: The PAL property is variable and can be updated during Unit Testing.

ReadingListKitTests has an example of this.

File: Sources/ReadingListKit/ReadingListModel.swift (hunk 9 above)
Comment (CMasterson) [reply Sources/ReadingListKit/ReadingListModel.swift]: Some older version of the code threw `using self before all properties are initialized` build errors, so I just sidestepped them.
Moving back to PAL init.

File: Demo/MozillaSocial-iOS/MoSoContent/MoSoContent.entitlements
Comment (CMasterson) [reply Demo/MozillaSocial-iOS/MoSoContent/MoSoContent.entitlements:1]: Removed

**Review by Gio2018 — DISMISSED** [review]
Hey @CMasterson , thanks for putting this together.
Before I do a full code review, I noticed a few things that require attention:

1. As already discussed in last week's meeting, there is an issue where, after authenticating, the reading list does not work as expected, and it still says that you should login to Pocket
2. If you relaunch you see the reading list.
3. If you go back to the user tab, despite the reading list being displayed, the signin/signup to Pocket appears again
4. If you sign in with a different user, you still see the old reading list

Also, if you tap on an item in the reading list, it does not open a web view (as described in the requirements document), but opens the related link in Safari.

Thank you!

**Comment by CMasterson** [comment]
> Hey @CMasterson , thanks for putting this together. Before I do a full code review, I noticed a few things that require attention:
> 
>     1. As already discussed in last week's meeting, there is an issue where, after authenticating, the reading list does not work as expected, and it still says that you should login to Pocket
> 
>     2. If you relaunch you see the reading list.
> 
>     3. If you go back to the user tab, despite the reading list being displayed, the signin/signup to Pocket appears again
> 
>     4. If you sign in with a different user, you still see the old reading list
> 
> 
> Also, if you tap on an item in the reading list, it does not open a web view (as described in the requirements document), but opens the related link in Safari.
> 
> Thank you!

Auth change updates are handled in a followup PR
https://github.com/MozillaSocial/mozilla-social-ios/pull/34

Webview VS Safari view will also be addressed in a followup pr, as we discussed. I don't feel like this should be a blocker to review.

File: Sources/ReadingListKit/ReadingListModel.swift (hunk 10 above)
Comment (CMasterson) [reply Sources/ReadingListKit/ReadingListModel.swift]: Does the backend not give us the items in the correct order?

Comment (dskuza) [reply Sources/ReadingListKit/ReadingListModel.swift]: I'm not positive… but it's best to verify what you're seeing on web or iOS vs what you're seeing in your code, just to double check. I remember sorting on iOS, but that might just be because we're fetching from Core Data and need to add a sort descriptor.

File: Sources/ReadingListKit/GraphQL/UnauthorizedInterceptor.swift (hunk 5 above)
Comment (dskuza) [reply Sources/ReadingListKit/GraphQL/UnauthorizedInterceptor.swift:46]: Keeping this around might be a "nice to have", since this is a possible scenario a user might encounter, and we'll need to account for it. Right now, though, it isn't used (outside of the Pocket codebase).

File: Sources/ReadingListKit/GraphQL/Apollo Extensions.swift (hunk 4 above)
Comment (dskuza) [reply Sources/ReadingListKit/GraphQL/Apollo Extensions.swift:24]: If we keep the `UnauthorizedErrorInterceptor`, then it's fine to keep this as-is.

File: Demo/MozillaSocial-iOS/apollo-ios-cli
Comment (dskuza) [reply Demo/MozillaSocial-iOS/apollo-ios-cli:1]: That might be possible. Not saying we should or shouldn't, but in the Pocket codebase, we do commit the generated files. We _might_ run into a thing where the build times increase a bit because we're potentially fetching and generating these files on ever build, unless maybe we add the `.graphQL` files as input files, and only run when they're changed.

Comment (CMasterson) [reply Demo/MozillaSocial-iOS/apollo-ios-cli:1]: I lean towards 'should' for this case.
People sometimes try edit the generated files to get what they want.
The files also add a lot of noise when we update our definitions and Glean has a timestamp in one of it's files that gets updated every build.

File: Sources/ReadingListKit/GraphQL/UnauthorizedInterceptor.swift (hunk 5 above)
Comment (CMasterson) [reply Sources/ReadingListKit/GraphQL/UnauthorizedInterceptor.swift:46]: I'm very much of the 'if it's not used, remove it' mindset here.
Would you object?

File: Sources/ReadingListKit/PocketAccessLayer.swift [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+
+public protocol ReadingListSession {
+    var token: String { get }
+    var guid: String { get }
+}
+
+protocol PocketAccessLayerProtocol {
+    var delegate: ReadingListModelDelegate? { get set }
+
+    func fetchSaves(after cursor: String?)
```
Comment (dskuza) [inline Sources/ReadingListKit/PocketAccessLayer.swift:19]: A welcome change!

File: Sources/ReadingListKit/PocketItem Wrapper.swift [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+protocol PocketItem {
+    var remoteID: String { get }
+    var title: String? { get }
+    var givenUrl: String { get }
+    var resolvedUrl: String? { get }
+    var image: String? { get }
+    var subtitle: String? { get }
+}
+
+extension PocketGraph.ItemByURLQuery.Data.ItemByUrl: PocketItem {
```
Comment (dskuza) [inline Sources/ReadingListKit/PocketItem Wrapper.swift:16]: I'm still unsure as to whether I think we should return _this_ type as a `PocketItem`, or have `PocketItem` itself be a struct. However, I think since both end options are value types, that this change is fine. Either way it's a good abstraction from GraphQL. I think I might've thought differently if Apollo was giving us back a reference type.

File: Sources/ReadingListKit/ReadingListModel.swift [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+        self.analytics = analyticsTracker
+        pocketAccessLayer = PocketAccessLayer(sessionProvider, consumerKey)
+        pocketAccessLayer.delegate = self
+
+        fetchMoreReadingList()
+    }
+
+    // MARK: - Fetch Reading List Items
+
+    func didDisplay(item: ReadingListCellViewModel) {
```
Comment (dskuza) [inline Sources/ReadingListKit/ReadingListModel.swift]: One thing that's throwing me off, here, is that we're using a view model as an identifier for an item, whereas in other places we're using a string (I believe its URL…?). The latter makes more sense to be, since the supplied view model here is a representation of what should be presented, and not necessarily an item itself. 

What do you think about updating this function to take in a `String` rather than a view model?

File: Sources/ReadingListKit/ReadingListModel.swift [hunk 19]
```diff
@@ -0,0 +1,154 @@
...
+        if allItemsAreDownloaded() {
+            return
+        }
+
+        let loadMoreThreshold = readingListItems.count - 3
+        if readingListItems.firstIndex(of: item) == loadMoreThreshold {
+            fetchMoreReadingList()
```
Comment (dskuza) [inline Sources/ReadingListKit/ReadingListModel.swift:53]: Clever!

File: Sources/ReadingListKit/ReadingListModel.swift [hunk 20]
```diff
@@ -0,0 +1,154 @@
...
+
+    func removeItemFromList(item: String) {
+        let filteredItems = self.readingListItems.filter {
+            $0.contentURL != item
+        }
+
+        readingListItems = filteredItems
+
+        if let oldTotal = totalNumberOfItemsInReadingList {
+            totalNumberOfItemsInReadingList = oldTotal - 1
```
Comment (dskuza) [inline Sources/ReadingListKit/ReadingListModel.swift:106]: Since we're using a `filter` above, we're assuming here that only one item should be removed. Is there a scenario where _more_ than one item could be removed, perhaps because our backend returned two items with the same `contentURL`? If that's the case, then we might be updating state here incorrectly. Just curious if that's something worth accounting for.

File: Sources/ReadingListKit/ReadingListModel.swift [hunk 21]
```diff
@@ -0,0 +1,154 @@
...
+    // MARK: - Archive Items
+
+    func archive(item: String) {
+        pocketAccessLayer.archive(item: item)
+    }
+
+    // MARK: - ReadingListItem Helpers
+
+    func readingListItem(from item: PocketItem) -> ReadingListCellViewModel {
+        let defaultImageURLString = "https://helios-i.mashable.com/imagery/articles/05fACELrEVc4kAfNQbhhcVh/hero-image.fill.size_1248x702.v1667556469.png"
```
Comment (dskuza) [inline Sources/ReadingListKit/ReadingListModel.swift]: I think we should just account for items without thumbnails, and not use a placeholder. But, that's a UX thing.

**Review by dskuza — COMMENTED** [review]
@CMasterson I think the changes definitely improved the overall code, so :+1: for that! I've left a few more comments that could be addressed, but don't necessarily _have_ to. Just food for thought. I think this is close to an approval, though.

File: Sources/ReadingListKit/GraphQL/Apollo Extensions.swift (hunk 2 above)
Comment (CMasterson) [reply Sources/ReadingListKit/GraphQL/Apollo Extensions.swift:17]: I did not intend to mark this as resolved.

File: Sources/ReadingListKit/ReadingListModel.swift (hunk 18 above)
Comment (CMasterson) [reply Sources/ReadingListKit/ReadingListModel.swift]: I agree that this is a bit clunky, the reason I did it was that it was easier to filter the main array by the VM rather than reaching into each to find it's URL.
But you're right, that is a better way and only slightly more effort.

File: Sources/ReadingListKit/ReadingListModel.swift (hunk 20 above)
Comment (CMasterson) [reply Sources/ReadingListKit/ReadingListModel.swift:106]: My understanding is that contentURL is the ID of the Save item. So it _shouldn't_ ever overlap.
Shouldn't......
I could add a check for multiple hits in the array and log that to Sentry or other relevant logging system.

File: Sources/ReadingListKit/ReadingListModel.swift (hunk 21 above)
Comment (CMasterson) [reply Sources/ReadingListKit/ReadingListModel.swift]: What do you mean account for them?
Is this a question for Design, do you think?

File: Sources/ReadingListKit/GraphQL/UnauthorizedInterceptor.swift (hunk 5 above)
Comment (dskuza) [reply Sources/ReadingListKit/GraphQL/UnauthorizedInterceptor.swift:46]: I like that mentality. It's an easy change to bring in once we have to account for it.

File: Sources/ReadingListKit/PocketAccessLayerImplementation.swift (hunk 7 above)
Comment (dskuza) [reply Sources/ReadingListKit/PocketAccessLayerImplementation.swift]: This is continued in another comment in the followup review.

File: Sources/ReadingListKit/ReadingListModel.swift (hunk 21 above)
Comment (dskuza) [reply Sources/ReadingListKit/ReadingListModel.swift]: If I'm interpreting the code correctly, I'm suggesting we _don't_ have a default thumbnail for items without a thumbnail URL.

File: Sources/ReadingListKit/GraphQL/Apollo Extensions.swift (hunk 2 above)
Comment (CMasterson) [reply Sources/ReadingListKit/GraphQL/Apollo Extensions.swift:17]: That function should never, ever, ever, ever return something that can't be rendered into a URL.
The Pocket code force unwrapped the created URL, I pulled that code into it's own fucntion so the effect is similar.

Do you still want to change this? If we don't have a URL we can't continue, so some sort of fatal error/exception seems appropriate.

File: Sources/ReadingListKit/GraphQL/Apollo Extensions.swift (hunk 4 above)
Comment (CMasterson) [reply Sources/ReadingListKit/GraphQL/Apollo Extensions.swift:24]: Looking deeper into this code; It's a little more complex than I realized. Ok to try address this in a followup PR? @dskuza

File: Sources/ReadingListKit/GraphQL/UnauthorizedInterceptor.swift (hunk 5 above)
Comment (CMasterson) [reply Sources/ReadingListKit/GraphQL/UnauthorizedInterceptor.swift:46]: As above, the surrounding code is more complex than I realized. This would also be caught in the followup PR, if you agree. @dskuza

File: Sources/ReadingListKit/PocketAccessLayerImplementation.swift (hunk 8 above)
Comment (CMasterson) [reply Sources/ReadingListKit/PocketAccessLayerImplementation.swift]: @dskuza Is the updated code what you expected?

File: Sources/ReadingListKit/ReadingListModel.swift (hunk 18 above)
Comment (CMasterson) [reply Sources/ReadingListKit/ReadingListModel.swift]: Updated!

File: Sources/ReadingListKit/ReadingListModel.swift (hunk 10 above)
Comment (CMasterson) [reply Sources/ReadingListKit/ReadingListModel.swift]: I checked and it seems like the pages come down in the expected order, but sometimes the items in the page are reordered (In one instance I think a page was reversed?).

Good enough for now?

File: Sources/ReadingListKit/PocketItem Wrapper.swift (hunk 17 above)
Comment (CMasterson) [reply Sources/ReadingListKit/PocketItem Wrapper.swift:16]: Both are fine options, this one just means I don't have to init a new instance of anything.

**Comment by CMasterson** [comment]
Simulator Screen Recording - iPad (10th generation) - 2023-12-11 at 10 53 58

**Review by dskuza — APPROVED** [review]
Nice, thanks for following up on all of our discussed changes. Onto the next piece of work!
